### DISCUSSION TITLE
Unable to Mock when running Pester in its own process/scope

### DISCUSSION THREAD

**Discussion by tbergstedt** [pester/Pester#2061] [Q&A] [answered]
As I am continuing to struggle with using Pester to test Pester commands, I have come upon a new problem. As described in [this issue](https://github.com/pester/Pester/issues/2058), I am now executing the Pester commands to be verified within its own scope, which works much better.

However, at a certain stage I need to be able to mock some commands that the functionality to be tested is using:
```powershell
#CommonFunctions.psm1
Function Get-SomeItem {
   # Retrieves data from net
}
Export-ModuleMember -Function Get-SomeItem
```
```powershell
#MockedFunctions.psm1
Function Get-MockedItem {
   # Mocks data retrieval from net
}
Export-ModuleMember -Function Get-MockedItem
```
The actual script:
```powershell
Describe "VerifyTests" {

  BeforeAll {
    $target = 'ScriptToVerify.ps1'

    $pesterModulePath = (Get-Module Pester).Path -replace '\.psm1','.psd1'
    $CommonFunctionsModulePath = "$PSScriptRoot\CommonFunctions.psm1"
    $MockedFunctionsModulePath = "$PSScriptRoot\MockedFunctions.psm1"
    
  }
      # Correct configurations
      It "ShouldPass" {

        # Check items with correct setup
        $correctRun = Start-Job {
          Import-Module $using:pesterModulePath
          Import-Module $using:CommonFunctionsModulePath 
          Import-Module $using:MockedFunctionsModulePath 

          Mock Get-SomeItem -ModuleName CommonModule -ParameterFilter { $Item = 1 } { Get-MockedItem 1 }

          $container = New-PesterContainer -Path "$using:target" -Data @{Item=1}
          Invoke-Pester -Container $container -PassThru -Output None
        } | Receive-Job -Wait

        $correctRun.TotalCount | Should -Be 1
        $correctRun.FailedCount | Should -Be 0
      }
    }
}
```
But when trying to execute this, the only result I get is `OperationStopped: Mock data are not setup for this scope, what happened?` I interpret this some user error, but I fail to understand how I shouldset this up? I have tried without `-ModuleName` and with using `InModuleScope`. It seems I am unable to use Mock at all from within a separate scope.

**Comment by fflaten** [comment, marked as answer]
You can't call `Mock` before `Invoke-Pester` in your job (which is a new process). Mocks need to be created during a Pester-run in the container where it's required. In other words, inside the `ScriptToVerify.ps1` testscript.

**Reply by tbergstedt** [reply]
I understand, but that won't work in my situation, since `ScriptToVerify.ps1` is the actual "production" script and it shouldn't contain any mocking. I'll experiment with wrapping the mocking + script execution.

**Reply by nohwnd** [reply]
What you should provide to Invoke-Pester is `ScriptToVerify.Tests.ps1` script, not `scriptToVerify.ps1`. As Frode pointed out the error happens because you are running `Mock` outside of a Pester run. Mock can only be used in a Pester run, inside of Describe/Context or It block.

How does the content of your `ScriptToVerify.ps1` looks like?

**Reply by tbergstedt** [reply]
The `ScriptToVerify.ps1` is simply a placeholder name for a number of scripts I have written that verifies some common rules and regulations regarding our code. I use Pester commands for executing them and for producing unit test summaries that can be published in Azure DevOps, etc. It's very neat to use Pester in this way.

The one I'm currently struggling with is ~500 lines, so I won't bother posting it all here, but what it basically does a number of times is looking up information within our source code that also should have corresponding activities in Azure DevOps. 

A source code file can have a notation that connects it to an issue/work item in Azure DevOps. The script then parses the information and checks the status of certain elements in it, like:
```powershell
Describe '<_> workitem readiness' -ForEach $itemsToCheck {
    $workItems = Get-WorkItemsFromFile $_
    foreach ($workitem in $workitems) {
        It 'Should have correct status' {
            $wi = Get-WorkItem $workitem
            $wi.Fields.AssignedTo | Should -Not -BeNullOrEmpty
            $wi.Fields.Status | Should -Not -BeIn @('New','Removed','Closed')
        }
    }
}
```
The command `Get-WorkItem` retrieves information from Azure DevOps, which can be prone to change. So when verifying changes to this script, I'd like to be able to Mock that function to instead return 'fake' work item info without actually contacting Azure DevOps and stays identical, etc. So that's basically what I was trying to do above.

I should say that I've progressed some in wrapping the mocking + Pester invocation into a temporary  script that then can be run in its own scope. Not there yet though.

**Reply by nohwnd** [reply]
Your mocks won't know where to bind. You'd have to invoke your tests in another Describe which you can do by passing the files as paths, and mocks around as scriptblock.

```powershell
param($Mocks, $TestFile)

Describe "wrapper" { 
    BeforeAll { 
        . $Mocks
     }
     
     . $TestFile
}
```

Something like this.

**Reply by tbergstedt** [reply]
Thanks, I'll try that. And I realize that part of my problem was that this particular script (the most advanced one) wasn't very well converted to Pester 5 notation, so it also boils down doing my homework...

Appreciate all your help!
